File: device-app/src/handlers/provision.rs
//! Development provisioning: hand a device a pressing it did not receive
//! through a ceremony.
//!
//! Staging a scene needs copies with plausible numbers, and the honest path to
//! copy #15 is fourteen earlier presses to fourteen distinct devices, because a
//! pressing binds to its recipient and a device holds one at a time. With two
//! Flex on a desk that is unreachable, so the relay acts as the master: it
//! holds the album key, signs the certificates, and pushes the result here.
//!
//! This does not weaken the model. Provisioning only ever *adds* a holding,
//! never removes one: a device that already holds a pressing is refused. The
//! certificates are verified exactly as `PRESS_ACCEPT` verifies them, including
//! that the bearer key handed in derives the `holderpub` the album key signed,
//! so nothing unsigned or mismatched can be planted. What is fictional is only
//! who pressed it.
//!
//! **This is a development path and its bearer key crosses the USB cable in
//! the clear.** There is no pairing here and so no session key to seal it with:
//! the relay is playing the master, and anything watching that cable owns the
//! copy. A real press seals the key against the paired channel; this one cannot
//! and must never be reached other than by a scene being staged on a desk.

use crate::certs::{parse_album_cert, parse_pressing_cert, ALBUM_CERT_LEN, PRESSING_CERT_LEN};
use crate::crypto;
use crate::handlers::press::BEARER_KEY_LEN;
use crate::session::Session;
use crate::state::Store;
use crate::AppSW;
use ledger_device_sdk::io::{Command, CommandResponse};

/// PROVISION_ALBUM: data = AlbumCert. Verified and staged for the pressing
/// step, which cannot carry both certificates in one 255-byte APDU. Uses the
/// session's staging slot but requires no pairing: there is no peer here, the
/// relay is the authority by construction.
pub fn handler_provision_album<'a>(
    command: Command<'a>,
    session: &mut Session,
) -> Result<CommandResponse<'a>, AppSW> {
    let data = command.get_data();
    if data.len() != ALBUM_CERT_LEN {
        return Err(AppSW::WrongApduLength);
    }
    parse_album_cert(data)?;
    session.staged_album.copy_from_slice(data);
    session.staged_album_valid = true;
    Ok(command.into_response())
}

/// PROVISION_PRESSING: data = PressingCert(178) || bearer key(32) = 210, one
/// frame. Same chain checks as PRESS_ACCEPT, then stores. Refuses to overwrite
/// an existing holding.
pub fn handler_provision_pressing<'a>(
    command: Command<'a>,
    session: &mut Session,
) -> Result<CommandResponse<'a>, AppSW> {
    if !session.staged_album_valid {
        return Err(AppSW::BadState);
    }
    let data = command.get_data();
    if data.len() != PRESSING_CERT_LEN + BEARER_KEY_LEN {
        return Err(AppSW::WrongApduLength);
    }
    let mut cert_buf = [0u8; PRESSING_CERT_LEN];
    cert_buf.copy_from_slice(&data[..PRESSING_CERT_LEN]);
    let mut bearer_priv = [0u8; BEARER_KEY_LEN];
    bearer_priv.copy_from_slice(&data[PRESSING_CERT_LEN..]);

    let staged_album = session.staged_album;
    let album = parse_album_cert(&staged_album)?;
    let pressing = parse_pressing_cert(&cert_buf, &album.albpub)?;

    let album_id = crypto::sha256(&[&album.albpub])?;
    if !crypto::mac_eq(&album_id, &pressing.album_id) {
        return Err(AppSW::BadCert);
    }
    if pressing.edition != album.edition {
        return Err(AppSW::BadCert);
    }
    if crypto::pubkey_of(&bearer_priv)? != pressing.holderpub {
        return Err(AppSW::BadCert);
    }
    let mut nvm = Store::get()?;
    if nvm.has_pressing == 1 {
        return Err(AppSW::BadState);
    }

    nvm.has_pressing = 1;
    nvm.pressing_cert = cert_buf;
    nvm.pressing_album_cert = staged_album;
    nvm.pressing_priv = bearer_priv;
    nvm.has_from = 0;
    nvm.ring_len = 0;
    Store::put(&nvm);
    session.staged_album_valid = false;

    Ok(command.into_response())
}

File: relay/provision.py
"""Development provisioning: hand a device a pressing without a ceremony.

A pressing binds to its recipient and a device holds one at a time, so the
honest route to copy #15 is fourteen earlier presses to fourteen distinct
devices. With two Flex on a desk that is unreachable, and staging a scene should
not depend on it. So the relay plays the master here: it derives the album key,
signs both certificates, carries the sleeve, and pushes the result.

The device still verifies everything it would verify on a real receive (both
signatures, the album_id linkage, the edition match, the binding to its own
public key) and still refuses to overwrite a pressing it already holds. Only the
authorship of the press is fictional.

    ./scripts/provision.sh --number 15 --edition 20
    ./scripts/provision.sh --device b --title "Discovery" --number 3 --edition 12

The album key is derived from title/artist/edition, so the Edition ID on screen
is stable across takes and across an NVM wipe: re-provisioning after a reflash
reproduces the same edition rather than a look-alike.
"""

import argparse
import hashlib
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "tests"))

from hid_device import HidDevice, enumerate_ledgers  # noqa: E402
from presse_client import (  # noqa: E402
    ART_LEN,
    SLOT_PRESSING,
    Presse,
    build_album_cert,
    build_pressing_cert,
    demo_album_key,
    demo_bearer_key,
    provision_pressing,
    upload_art,
    verify_chain,
)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--device", choices=["a", "b"], default="a",
                        help="which HID path to provision (a = paths[0])")
    parser.add_argument("--title", default="Random Access Memories")
    parser.add_argument("--artist", default="Daft Punk")
    parser.add_argument("--edition", type=int, default=20)
    parser.add_argument("--number", type=int, default=15)
    parser.add_argument("--cover", default="ram-cover.bin",
                        help="sleeve in docs/art/, sealed into the album cert")
    args = parser.parse_args()

    if not 0 < args.number <= args.edition:
        sys.exit(f"number {args.number} outside an edition of {args.edition}")

    paths = enumerate_ledgers()
    idx = 0 if args.device == "a" else 1
    if len(paths) <= idx:
        sys.exit(f"device {args.device} not found ({len(paths)} Flex attached)")
    dev = Presse(HidDevice(f"Flex {args.device.upper()}", paths[idx]))

    info = dev.get_info()
    if info["has_pressing"]:
        sys.exit("this device already holds a pressing; provisioning never overwrites one")

    cover_path = os.path.join(os.path.dirname(__file__), "..", "docs", "art", args.cover)
    with open(cover_path, "rb") as fh:
        art = fh.read()
    if len(art) != ART_LEN:
        sys.exit(f"{args.cover} is {len(art)} bytes, the device expects {ART_LEN}")
    sleeve_hash = hashlib.sha256(art).digest()

    priv, albpub = demo_album_key(args.title, args.artist, args.edition)
    album_cert = build_album_cert(
        priv, albpub, args.title, args.edition, sleeve_hash, args.artist
    )
    album_id = hashlib.sha256(albpub).digest()
    # The copy is bound to a bearer key, not to this device: provisioned or
    # pressed, a copy is transferable afterwards and nothing here should make
    # it less so. The relay mints that key and hands over both halves.
    bearer_priv, bearer_pub = demo_bearer_key(
        args.title, args.artist, args.edition, args.number
    )
    pressing_cert = build_pressing_cert(
        priv, album_id, args.number, args.edition, bearer_pub
    )

    # Verified here too, against the same rules the device applies: a relay bug
    # should fail on the laptop, not as an opaque status word on the device.
    result = verify_chain(album_cert, pressing_cert)
    assert result["holderpub"] == bearer_pub, "bearer key does not match the certificate"

    print(f'== provisioning Flex {args.device.upper()} ==')
    print(f'   "{args.title}" by {args.artist}, #{args.number} of {args.edition}')
    upload_art(dev, art, SLOT_PRESSING)
    print(f"   sleeve carried into the pressing slot ({args.cover}, {len(art)} bytes)")
    provision_pressing(dev, album_cert, pressing_cert, bearer_priv)
    print(f"   edition ID {hashlib.sha256(albpub).hexdigest()[:8].upper()}")
    print("   provisioned. The device verified both signatures and the bearer key.")


if __name__ == "__main__":
    main()
